src/query/storage/m3/storage.go

```diff
@@ -361,3 +354,150 @@ func (s *m3storage) writeSingle(
+
+	if len(r.completeAggregated) == 0 {
+		// Absolutely no complete aggregated namespaces, need to fanout to all
+		// partial aggregated namespaces as well as the unaggregated cluster
+		// as we have no idea who has the longest retention
+		result := append(r.partialAggregated, unaggregated)
+		return namespaceCoversPartialQueryRange, result, nil
+	}
+
+	// Return the longest retention aggregated namespace and
+	// any potentially more granular or longer retention partial
+	// aggregated namespaces
+	sort.Stable(sort.Reverse(ClusterNamespacesByRetentionAsc(r.completeAggregated)))
+
+	// Take longest retention complete aggregated namespace or the unaggregated
+	// cluster if that is longer than the longest aggregated namespace
+	result := r.completeAggregated[:1]
+	completedAttrs := result[0].Options().Attributes()
+	if completedAttrs.Retention <= unaggregatedRetention {
+		// If the longest aggregated cluster for some reason has lower retention
+		// than the unaggregated cluster then we prefer the unaggregated cluster
+		// as it has a complete data set and is always the most granular
+		result[0] = unaggregated
+		completedAttrs = unaggregated.Options().Attributes()
+	}
+
+	// Take any partially aggregated namespaces with longer retention or
+	// same retention with more granular resolution that may contain
+	// a matching metric
+	for _, n := range r.partialAggregated {
+		if n.Options().Attributes().Retention > completedAttrs.Retention {
+			// Higher retention
+			result = append(result, n)
+			continue
+		}
+		if n.Options().Attributes().Retention == completedAttrs.Retention &&
+			n.Options().Attributes().Resolution < completedAttrs.Resolution {
+			// Same retention but more granular resolution
+			result = append(result, n)
+			continue
+		}
+	}
+
+	return namespaceCoversPartialQueryRange, result, nil
+}
+
+type reusedAggregatedNamespaceSlices struct {
+	completeAggregated []ClusterNamespace
+	partialAggregated  []ClusterNamespace
+}
+
+func (s *m3storage) aggregatedNamespaces(
+	slices reusedAggregatedNamespaceSlices,
+	filter func(ClusterNamespace) bool,
+) reusedAggregatedNamespaceSlices {
+	all := s.clusters.ClusterNamespaces()
+
+	// Reset reused slices as necessary
+	if slices.completeAggregated == nil {
+		slices.completeAggregated = make([]ClusterNamespace, 0, len(all))
+	}
+	slices.completeAggregated = slices.completeAggregated[:0]
+	if slices.partialAggregated == nil {
+		slices.partialAggregated = make([]ClusterNamespace, 0, len(all))
+	}
+	slices.partialAggregated = slices.partialAggregated[:0]
+
+	for _, namespace := range all {
+		opts := namespace.Options()
+		if opts.Attributes().MetricsType != storage.AggregatedMetricsType {
+			// Not an aggregated cluster
+			continue
+		}
+
+		if filter != nil && !filter(namespace) {
+			continue
+		}
+
+		downsampleOpts, err := opts.DownsampleOptions()
+		if err != nil || !downsampleOpts.All {
+			// Cluster does not contain all data, include as part of fan out
```

Comment:
Confused about the downsampling thing, how is this different than saying something is unaggregated?

Reply:
Basically it's aggregated, but data will only go to this cluster if rules are setup to do so.